snuba/datasets/errors_replacer.py

```diff
@@ -242,6 +244,100 @@ def _build_event_tombstone_replacement(
     )
 
 
+def _build_event_merge_replacement(
+    message: Mapping[str, Any],
+    where: str,
+    query_args: Mapping[str, str],
+    query_time_flags: Tuple[Any, ...],
+    all_columns: Sequence[FlattenedColumn],
+) -> Optional[Replacement]:
+    # HACK: We were sending duplicates of the `end_merge` message from Sentry,
+    # this is only for performance of the backlog.
+    txn = message.get("transaction_id")
+    if txn:
+        if txn in SEEN_MERGE_TXN_CACHE:
+            return None
+        else:
+            SEEN_MERGE_TXN_CACHE.append(txn)
+
+    timestamp = datetime.strptime(message["datetime"], settings.PAYLOAD_DATETIME_FORMAT)
+    all_column_names = [c.escaped for c in all_columns]
+    select_columns = map(
+        lambda i: i if i != "group_id" else str(message["new_group_id"]),
+        all_column_names,
+    )
+
+    count_query_template = (
+        """\
+        SELECT count()
+        FROM %(table_name)s FINAL
+    """
+        + where
+    )
+
+    insert_query_template = (
+        """\
+        INSERT INTO %(table_name)s (%(all_columns)s)
+        SELECT %(select_columns)s
+        FROM %(table_name)s FINAL
+    """
+        + where
+    )
+
+    final_query_args = {
+        "all_columns": ", ".join(all_column_names),
+        "select_columns": ", ".join(select_columns),
+        "project_id": message["project_id"],
+        "timestamp": timestamp.strftime(DATETIME_FORMAT),
+    }
+    final_query_args.update(query_args)
+
+    return Replacement(
+        count_query_template, insert_query_template, final_query_args, query_time_flags
+    )
+
+
+def process_merge_events(
+    message: Mapping[str, Any], all_columns: Sequence[FlattenedColumn]
+) -> Optional[Replacement]:
+    """
+    Merge individual events into new group. The old group will have to be
+    manually excluded from search queries.
+
+    See docstring of process_exclude_groups for an explanation of how this is used.
+
+    Note that events merged this way cannot be cleanly unmerged by
+    process_unmerge, as their group hashes possibly stand in no correlation to
+    how the merging was done.
+    """
+
+    event_ids = message["event_ids"]
+    if not event_ids:
+        return None
+
+    # XXX: We need to construct a query that works on both event_id columns,
+    # either represented as UUID or as hyphenless FixedString. That's why we
+    # use replaceAll(toString()).
+    where = """\
```

Comment:
@lynnagara I know you've been working to align all the types of the different tables, is the event_id still different between the tables?

Reply:
Yes, it's UUID in errors, and hyphen less fixed string in events. It'll stay different until the events storage eventually goes away. Isn't `replaceAll(toString(event_id), '-', '')` in prewhere going to cause performance issues since it will be run on every single row? It may be better to pass the state_name into this function (so we know whether we are dealing with the events or errors table) then format the list of event_ids separately for each storage so the `replaceAll(toString(...))` part is not needed.